### PR TITLE
*: bump prometheus and fix fallout

### PR DESCRIPTION
**Pull request by MichaHoffmann**
* [x] I added CHANGELOG entry for this change.
* [x] Change is not relevant to the end user.

## Changes

* bump prometheus to latest main
* fix much fallout

## Verification

* rely on previous tests

## Attribution

* most of the work was done by @yeya24

### CODE REVIEW

File: pkg/query/querier.go [hunk 1]
```diff
@@ -439,23 +430,23 @@ func (q *querier) LabelValues(name string, matchers ...*labels.Matcher) ([]strin
 		return nil, nil, errors.Wrap(err, "proxy LabelValues()")
 	}
 
-	var warns storage.Warnings
+	var warns annotations.Annotations
 	for _, w := range resp.Warnings {
-		warns = append(warns, errors.New(w))
+		warns.Add(errors.New(w))
 	}
```
Comment (yeya24) [inline pkg/query/querier.go:436]: Can also just use `warns := annotations.New().Merge(resp.warnings)`

File: pkg/query/querier.go [hunk 2]
```diff
@@ -472,15 +463,14 @@ func (q *querier) LabelNames(matchers ...*labels.Matcher) ([]string, storage.War
 		return nil, nil, errors.Wrap(err, "proxy LabelNames()")
 	}
 
-	var warns storage.Warnings
+	var warns annotations.Annotations
 	for _, w := range resp.Warnings {
-		warns = append(warns, errors.New(w))
+		warns.Add(errors.New(w))
 	}
```
Comment (yeya24) [inline pkg/query/querier.go:469]: ditto

File: pkg/rules/queryable.go [hunk 3]
```diff
@@ -81,7 +81,7 @@ func (q *promClientsQueryable) Querier(ctx context.Context, mint, maxt int64) (s
 }
 
 // Select implements storage.Querier interface.
-func (q *promClientsQuerier) Select(_ bool, _ *storage.SelectHints, matchers ...*labels.Matcher) storage.SeriesSet {
+func (q *promClientsQuerier) Select(ctx context.Context, _ bool, _ *storage.SelectHints, matchers ...*labels.Matcher) storage.SeriesSet {
```
Comment (yeya24) [inline pkg/rules/queryable.go:84]: Since we have `ctx` here, we should change L91 to use this ctx

Comment (MichaHoffmann) [reply pkg/rules/queryable.go:84]: :heavy_check_mark:

File: pkg/query/querier.go (hunk 1 above)
Comment (MichaHoffmann) [reply pkg/query/querier.go:436]: No, storepb APIs return []string for warnings still

File: pkg/query/querier.go (hunk 2 above)
Comment (MichaHoffmann) [reply pkg/query/querier.go:469]: same as above

**Comment by yeya24** [comment]
Overall, this looks great! I retried the CI failure and we can merge if succeeds

**Comment by yeya24** [comment]
```
=== NAME  TestStoreGateway
    store_gateway_test.go:117: store_gateway_test.go:117: ""
        
         unexpected error: block creation: add sample: 20 observations found in buckets, but the Count field is 9: histogram's observation count should be at least the number of observations found in the buckets
```

**Comment by MichaHoffmann** [comment]
> ```
> === NAME  TestStoreGateway
>     store_gateway_test.go:117: store_gateway_test.go:117: ""
>         
>          unexpected error: block creation: add sample: 20 observations found in buckets, but the Count field is 9: histogram's observation count should be at least the number of observations found in the buckets
> ```

So the reason is that we try to write a malformed histogram and tsdb appends now have validation for that.

File: pkg/block/indexheader/header_test.go [hunk 4]
```diff
@@ -519,78 +521,3 @@ func readSymbols(bs index.ByteSlice, version, off int) ([]string, map[uint32]str
 	}
 	return symbolSlice, symbols, errors.Wrap(d.Err(), "read symbols")
 }
-
-func TestIndexHeaderV1LookupSymbols(t *testing.T) {
```
Comment (fpetkovski) [inline pkg/block/indexheader/header_test.go:523]: Is this for the old TSDB 1 format?

**Review by fpetkovski — APPROVED** [review]

File: pkg/block/indexheader/header_test.go (hunk 4 above)
Comment (MichaHoffmann) [reply pkg/block/indexheader/header_test.go:523]: I think yes, we added it recently but having Symbols as an interface just for this one test isnt much bang for the buck i figured.

**Comment by yeya24** [comment]
This seems reproducible every time.

```
=== NAME  TestStoreGatewayLazyExpandedPostingsEnabled
    store_gateway_test.go:1215: store_gateway_test.go:1215: ""
        
         unexpected error: unable to find metrics [thanos_bucket_store_lazy_expanded_postings_total] with expected values after 50 retries. Last error: <nil>. Last values: [0]
        
```

**Comment by yeya24** [comment]
Nice work! Merging now

**Review by yeya24 — APPROVED** [review]

**Pull request merged**